test/unit/lib/plugins/aws/customResources/generateZip.test.js

```diff
@@ -0,0 +1,279 @@
+          await generate(cachePath);
+          return cachePath;
+        });
+      });
+
+      it('should write a file', async () => {
+        await generateZip(artefactName);
+
+        const exists = fse.existsSync(cachedFile);
+        expect(exists).to.be.true;
+        expect(renameStub).to.have.been.calledOnce;
+      });
+    });
+
+    describe('when the file already exists by the time the zip is generated', () => {
+      beforeEach(async () => {
+        // Override the ensureArtifact function so that it always calls the 'generate' function
+        // even if the file already exists
+        // This will simulate the scenario when multiple serverless instances are packaging at once
+        // and one of the instances writes to the cache path while the others are still generating the zip file
+        ensureArtifactStub.callsFake(async (_, generate) => {
+          await generate(cachePath);
+          return cachePath;
+        });
+
+        // Create an existing file that should conflict with the generated artefact
+        fse.ensureFileSync(cachedFile);
+        fse.writeFileSync(cachedFile, 'cached data');
+      });
+
+      describe('when the cached file is open for writing', () => {
+        it('should overwrite the cached file', async () => {
+          const fd = fs.openSync(cachedFile, 'r+');
+
+          try {
+            await generateZip(artefactName);
+
+            const cachedData = fse.readFileSync(cachedFile).toString();
+            expect(cachedData).not.to.eq('cached data');
+            expect(renameStub).to.have.been.calledOnce;
+          } finally {
+            fs.closeSync(fd);
+          }
+        });
+      });
+
+      describe('when the cached file is not open', () => {
+        it('should overwrite the cached file with a valid zip file', async () => {
+          await generateZip(artefactName);
+
+          // Check that the file was actually overwritten
+          const cachedData = fse.readFileSync(cachedFile).toString();
+          expect(cachedData).not.to.eq('cached data');
+
+          expect(renameStub).to.have.been.calledOnce;
+
+          // List the files in the zip to make sure it is valid
+          const filesInZip = await listZipFiles(cachedFile);
+          expect(filesInZip).to.have.all.members([
+            'README.md',
+            'apiGatewayCloudWatchRole/handler.js',
+            'cognitoUserPool/handler.js',
+            'cognitoUserPool/lib/permissions.js',
+            'cognitoUserPool/lib/userPool.js',
+            'eventBridge/handler.js',
+            'eventBridge/lib/eventBridge.js',
+            'eventBridge/lib/permissions.js',
+            'eventBridge/lib/utils.js',
+            's3/handler.js',
+            's3/lib/bucket.js',
+            's3/lib/permissions.js',
+            'utils.js',
+          ]);
+        });
+      });
+    });
+  });
+
+  describe('when the temporary directory is on a different device to the cache directory', () => {
```

Comment:
Can we cover those two scenarios (1) when tmp dir is on different device (2) when tmp dir is not on different device, with reusable group of tests?

What I mean is something as:

```javascript
const testMove = (beforeEach) => {
    beforeEach(() => beforeEach());
    describe('when the file does not exist by the time the zip is generated, ... );
    ..
});

describe('when the temporary directory is not on a different device to the cache directory', () => {
  testMove(() => ...);
});

describe('when the temporary directory is on a different device to the cache directory', () => {
  testMove(() => ...);
});
```

Reply:
Oh of course! That makes sense. I'll make the change

Reply:
In my last commit where I extracted the saveMoveFile tests I factored out the test scenarios like this